### PR TITLE
pardot-oauth

### PR DESCRIPTION
**Pull request by jacobrobertbaca**
# Description of change
got requests working with oauth flow

# Manual QA steps
 - Test creating a new connection with oauth
   - Works fine
 - Test creating a new connection but fail oauth
    - Fails correctly.
 - Tested creating a new connection with old auth, fake creds
    - Correctly got failure to login error
    - Note: this is where we expect old connections to fail after feb 15
 - Tested reauthing an old auth connection with oauth
    - Works after reauthing
 - Tested cloned old auth connection, real creds
    - correctly works as usual
# Risks
 - moderate
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_pardot/client.py [hunk 1]
```diff
@@ -41,7 +57,20 @@ class Client:
...
     def __init__(self, creds):
         self.creds = creds
-        self.login()
+        self.api_version = "4"
+        if self.has_oauth_values():
+            self.refresh_credentials()
+        elif self.has_api_key_auth_values():
+            self.login()
+        else:
+            raise AuthCredsMissingError("Requires OAuth credentials.")
```
Comment (asaf-erlich) [inline tap_pardot/client.py]: Shouldn't this say requires oauth credentials or "email, api key, and password" or whatever the other cred config values were?

File: tap_pardot/client.py [hunk 2]
```diff
@@ -145,7 +227,7 @@ def describe(self, endpoint, **kwargs):
         jitter=None,
     )
     def _fetch(self, method, endpoint, format_params, **kwargs):
-        base_formatting = [endpoint, self.api_version]
+        base_formatting = [endpoint, '{}']
```
Comment (asaf-erlich) [inline tap_pardot/client.py:232]: This looks really weird, why are you calling format here, putting a thing that will be later formatted again with the api version?

File: tap_pardot/client.py [hunk 3]
```diff
@@ -118,17 +197,20 @@ def _make_request(self, method, url, params=None):
                     method, url, headers=self._get_auth_header(), params=params
                 )
                 content = response.json()
+            if error_code == 89:
+                LOGGER.info("Pardot returned error code 89, switching to api version 3")
+                self.api_version = "3"
```
Comment (asaf-erlich) [inline tap_pardot/client.py:204]: Please link to pardot documentation here on why this is the right way to handle this error

File: tap_pardot/client.py [hunk 4]
```diff
@@ -77,24 +107,73 @@ def _check_error(self, content, activity):
...
         )
+
         response = requests.request(
-            method, url, headers=self._get_auth_header(), params=params
+            method, url.format(self.api_version), headers=self._get_auth_header(), params=params
         )
 
+        if response.status_code == 401:
+            if self.has_oauth_values():
+                LOGGER.info("Received a 401 unauthenticated error from Pardot. Reauthing and retrying the request.")
```
Comment (asaf-erlich) [inline tap_pardot/client.py]: I might emit this as a warning, not an info level... what do you think?

**Review by asaf-erlich — CHANGES_REQUESTED** [review]
Made comments above

File: tap_pardot/client.py (hunk 1 above)
Comment (zachharris1) [reply tap_pardot/client.py]: We originally said something along the lines of "Requires api key auth or oauth auth credentials (you have neither)" but this message treats api key auth more like a deprecated feature, which is closer to the truth.

File: tap_pardot/client.py (hunk 3 above)
Comment (zachharris1) [reply tap_pardot/client.py:204]: https://developer.pardot.com/kb/error-codes-messages/#error-code-89

File: tap_pardot/client.py (hunk 4 above)
Comment (zachharris1) [reply tap_pardot/client.py]: I think warning is appropriate if we expect all (most?) tap runs to last less than the length of time it takes an access token to expire, but we've yet to figure out how long that is.

File: tap_pardot/client.py (hunk 2 above)
Comment (zachharris1) [reply tap_pardot/client.py:232]: With old auth, the appropriate api version is determined on login. As far as we know, oauth doesn't provide a way to figure out what api version to use. Our solution is to guess that the api is version 4 in the init, then change the api version to 3 if we catch an 89 error in requests. Passing '{}' lets us format and add the api version after we get an 89.

File: tap_pardot/client.py (hunk 1 above)
Comment (asaf-erlich) [reply tap_pardot/client.py]: Sounds good, can you specify the config values mentioned for oauth then?

File: tap_pardot/client.py (hunk 2 above)
Comment (asaf-erlich) [reply tap_pardot/client.py:232]: `[endpoint, '{api_version}'] is an option and then in the second format use the keyword .format method: https://realpython.com/python-formatted-output/

**Review by asaf-erlich — APPROVED** [review]

**Review by KAllan357 — APPROVED** [review]

**Pull request merged**